mhs/common/mhs_common/messages/ebxml_request_envelope.py

```diff
@@ -107,19 +186,67 @@ def _extract_message_parts(msg: email.message.Message) -> Tuple[str, str]:
         # (if present) must be the first additional attachment.
 
         if not msg.is_multipart():
-            raise ebxml_envelope.EbXmlParsingError("Non-multipart message received!")
+            logger.error('0005', 'Non-multipart message received')
+            raise ebxml_envelope.EbXmlParsingError("Non-multipart message received")
 
-        message_parts = msg.get_payload()
+        message_parts: Sequence[email.message.EmailMessage] = tuple(msg.iter_parts())
 
         for i, part in enumerate(message_parts):
             if part.defects:
-                logger.warning('0003', 'Found defects in {PartIndex} of MIME message during parsing. {Defects}',
+                logger.warning('0006', 'Found defects in {PartIndex} of MIME message during parsing. {Defects}',
                                {'PartIndex': i, 'Defects': part.defects})
 
-        ebxml_part = message_parts[0].get_payload()
+        ebxml_part, is_base64_ebxml_part = EbxmlRequestEnvelope._convert_message_part_to_str(message_parts[0])
+        if is_base64_ebxml_part:
+            logger.error('0007', 'Failed to decode ebXML header part of message as text')
+            raise ebxml_envelope.EbXmlParsingError("Failed to decode ebXML header part of message as text")
 
         payload_part = None
+        attachments = []
         if len(message_parts) > 1:
-            payload_part = message_parts[1].get_payload()
+            payload_part, is_base64_payload = EbxmlRequestEnvelope._convert_message_part_to_str(message_parts[1])
+            if is_base64_payload:
+                logger.error('0008', 'Failed to decode HL7 payload part of message as text')
+                raise ebxml_envelope.EbXmlParsingError("Failed to decode HL7 payload part of message as text")
+
+            for attachment_message in message_parts[2:]:
```

Comment:
TODO-maybe bung a comment above this to make it a bit clearer as to the flow of this function, maybe throughout put some comments. Or do as a separate method.